### DISCUSSION TITLE
GSoC and the Asymmetric Travelling Salesperson Problem

### DISCUSSION THREAD

**Discussion by mjschwenne** [networkx/networkx#4705] [Q&A] [answered]
Hello everyone!

My name is Matt Schwennesen and I am a student at Michigan Technological University looking to participate in the 2021 Google Summer of Code.

So far I have read the GitHub pull requests relating to the issue and the research paper linked on the mentored projects page for the directed travelling salesperson problem.  I also read though the code in the pull request. This problem is both very interesting from a theoretical sense and a practical one as for example flights may have different costs or flight times depending on the direction.

It seems to me that the key to this algorithm is finding an approximation of the minimum spanning tree in the directed graph, which could actually have applications elsewhere in the project as well, seeing that none of the algorithms of `mst.py` in `netwrokx.algorithms.tree` work with a directed graph.  

Once the spanning tree is found, we should be able to use `networkx.algorithms.flow.network_simplex` to find the minimum flow in the graph, which can be shortcut to find the Hamiltonian cycle.

I have not seen any other pieces of the NetworkX code which could be used to facilitate this problem, but I have not conducted a deep search of the source code to see if there are any more pieces I could use. I am happy to discuss and accept any recommendations.

**Comment by dschult** [comment, marked as answer]
I'm not sure what a good definition of minimum spanning tree is for a directed graph.
Probably the best approach is that from the article from [Asadpour](https://pubsonline.informs.org/doi/pdf/10.1287/opre.2017.1603), (linked in the [projects page](https://networkx.org/documentation/latest/developer/projects.html))
or possibly from more recent papers that cite that paper.

**Reply by mjschwenne** [reply]
Right, the idea of a minimum spanning tree is difficult and poorly defined for a directed graph. Asadpour is using lambda-random spanning trees, with an additional set of restrictions which allow it to be used similarly to the minimum spanning tree in the Christofides algorithm. My original intend with mentioning the other minimum spanning tree algorithms in the project was that this component of the Asadpour algorithm may have uses elsewhere as it is an approximate minimum of the random spanning trees.

**Reply by dschult** [reply]
I agree that the lambda-random spanning trees approach would be helpful in other algorithms too. There's a set of algorithms aiming for low-stretch spanning trees which could build off of that for example.  Other examples too I'm sure. It'd be very useful.

**Comment by hannahtro** [comment]
Hello,
I am interested in the project. I have worked with networkx before and am very interested in graph algorithms. What is expected of the application?

**Reply by MridulS** [reply]
Hi @hannahtro!

To get acquainted with our development process please go through https://networkx.org/documentation/latest/developer/contribute.html#dev-workflow to setup
your local dev environment and https://networkx.org/documentation/latest/developer/new_contributor_faq.html for
some common FAQs.

For some issues to work on to get familiarised with NetworkX you can look at
https://github.com/networkx/networkx/issues?q=is%3Aissue+is%3Aopen+label%3ASprint or you can pick any
other issue on the GitHub issue tracker https://github.com/networkx/networkx/issues.

Please go through all the details available at https://summerofcode.withgoogle.com regarding application process.
You can look at a sample template at https://github.com/numfocus/gsoc/blob/master/templates/proposal.md

**Comment by mjschwenne** [comment]
Hey all!

While I have not been able to work on this as much as I would have liked (between course finals and a move across state coming up), I have been able to write another [blog post](https://mjschwenne.github.io/2021/05/08/held-karp-separation-oracle.html) about the algorithms underlying the Asadpour atsp algorithm, this time about building a separation for the ellipsoid algorithm. 

Thanks
